### ISSUE TITLE
Get length info of nodes and comment handling right

### ISSUE THREAD

**Issue by Up2Go-DavidRenz** [Up2Go/pmd#23]
Issues created by some rules* don't get the right begin or end lines or don't get assigned to the related method/class, so we can't see the origin of the issue. VariableNamingConventionsRule seems to find them.

- [ ] MethodNamingConventionsRule
- [ ] StdCyclomaticComplexityRule
- [ ] ExcessivePublicCountRule
- [ ] ExcessiveParameterListRule
- [ ] ExcessiveMethodLengthRule
- [ ] ClassNamingConventionsRule

Related JSON-report snippet:
```
{  
   "type":"issue",
   "check_name":"ExcessiveParameterList",
   "description":"Avoid long parameter lists.",
   "content":{  
      "body":" Methods with numerous parameters are a challenge to maintain, especially if most of them share the same datatype. These situations usually denote the need for new objects to wrap the numerous parameters.    "
   },
   "categories":[  
      "Style"
   ],
   "location":{  
      "path":"/src/fflib_SObjectDomain.cls",
      "lines":{  
         "begin":1,
         "end":977
      }
   },
   "severity":"normal"
}
```

or 

```
{  
   "type":"issue",
   "check_name":"ClassNamingConventions",
   "description":"Class names should begin with an uppercase character",
   "content":{  
      "body":" Class names should always begin with an upper case character.       "
   },
   "categories":[  
      "Style"
   ],
   "location":{  
      "path":"/src/fflib_SObjectDomain.cls",
      "lines":{  
         "begin":1,
         "end":977
      }
   },
   "severity":"critical"
}
```

**Comment by up2go-rsoesemann** [comment]
I think we are getting into bigger problems with linenumbers in general.

If the Jorje parser is taking out comments from the AST like @vazexqi said the line number code @adangel has contributed in https://github.com/Up2Go/pmd/commit/46c24f3a859a6df586e58114cb3d2553224a95be is never able to provide correct start and end numbers.

But I guess Code Climate is dependent on us giving back correct line numbers to show issues in Github, right @mrb ?!

**Comment by up2go-rsoesemann** [comment]
@vazexqi Couldn't the [`ParserLocationTranslator`](https://github.com/forcedotcom/idecore/blob/27eb24b00bfe844677f7945ec626bcbb7e2d563f/com.salesforce.ide.ui.editors/src/com/salesforce/ide/ui/editors/apex/util/ParserLocationTranslator.java) in your repo provide a solution to this?

**Comment by vazexqi** [comment]
That's specific to Eclipse. It's for highlighting the code in the editor based on your selection.

@up2go-rsoesemann - I think you misunderstood what 
I said about taking out comments. Comments are not "taken out". They are parsed into a separate [channel](https://theantlrguy.atlassian.net/wiki/pages/viewpage.action?pageId=2687083) by Antlr and there are not directly accessible using the AST. They do *not* change the line number.

```
1: // Some comment
2: public class MyClass {}
```

MyClass is still on line 2. It is **not** going to be shifted to line 1, as you seem to have assumed.

So, I think there is some other issue at hand. You are probably not giving PMD what it wants as the start and endline. As commit 46c24f3 seems to indicate, maybe you want the whole range of the method rather than the start line of the method name?

**Comment by up2go-rsoesemann** [comment]
@vazexqi could you please elaborate what I should change. @adangel's fix was better than what we had before but he doesn't know the semantics of the Jorje RealLoc as we don't do. Could you provide a code example on how to get the exact range (method going from and to) or make the fix in the code yourself?

**Comment by vazexqi** [comment]
Can you give me the exact setup steps that I need to reproduce this issue? I think it's just a test run but can you just tell me which test to run? Thanks!

**Comment by up2go-rsoesemann** [comment]
@vazexqi Instead of reproducing in a test I think it's better to show you how to open the Rule Designer:

1. Locally clone this repo. You should have a bunch of `pmd-xyz` folders in your git folder after that
2. Go into the `/pmd/pmd-apex/` folder
3. Build and run the Rule Designer GUI with this command `mvn -P designer exec:java`
4. Play around with AST viewer (add comments to class and methods and see how line comments are not taken out.

![startendline](https://cloud.githubusercontent.com/assets/8180281/14348483/0a07686c-fcbd-11e5-86d0-5b3bbfc60de7.png)

The start and endline is calculated here using `RealLoc`:

https://github.com/Up2Go/pmd/blob/master/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/AbstractApexNode.java#L42

We have a small and a bigger problem with that:

- SMALL PROBLEM: Some rules reject to long methods or classes. As we calculate both by simply doing a `endline -startLine` including comments we penalize the wrong things.
- BIG PROBLEM: We pass start and endline to the Code Climate UI which is highlighting the problematic section of code pushed into GitHub. If we have offsets there people will "hate us" ;-)

![](http://blog.codeclimate.com/images/posts/javascript/js_code_detail.png)

_If you need help getting the designer to run @Up2Go-DavidRenz or myself can guide you in a Skype chat_.

**Comment by up2go-rsoesemann** [comment]
Improved this in https://github.com/Up2Go/pmd/commit/5e9df4a4468c3001b1ca1c70baee930ab0855535. Class now don't always start in line 1.

**Comment by Up2Go-DavidRenz** [comment]
Inner classes get the same line information as their outer classes, so if the outer class hits the excessive class length limit all inner classes also throw an issue. For example the [MetadataService.cls](https://github.com/Up2Go/pmd/blob/master/pmd-apex/src/test/resources/MetadataService.cls) creates round about 700 excessive class length issues.

**Comment by vazexqi** [comment]
@up2go-rsoesemann Thanks for the example. I see the issue now with the size.

The fundamental issue is that you are trying to get the size of the method/class by doing endLine - startLine. That, will of course, take into consideration all the comments since startLine is the *exact* location of the element in the source code i.e., what line it appears in your file, while must consider offsets from comments.

We cannot change the semantics of endLine or startLine. 

Instead, we need to change the way you determine class/method size. Of the top of my head, what I would do is iterate through each statement in the class and add up the sum of statements on different lines. That will give you a good estimate for size and will be the exact thing as thought you had all comments removed. The downside is that you need more work to calculate size.

Of course, you can fake this by, putting all statements in one line, and then your rule won't work. But, again, this is also a trick that people can use even if you remove the comments and use your original method.

That's way, people have come up with more complex metrics for estimating complexity such as number of methods, number of (accessible, inheritable) fields, and cyclomatic complexity. Those are not fooled by use of code formatting.



By the way, I still don't quite understand what you mean by the BIG PROBLEM with CodeClimate. What is needed to solve the problem there? Remember I don't use CodeClimate and I am not familiar with their code.

**Comment by up2go-rsoesemann** [comment]
@vazexqi Thanks for answer which indeed is very helpful. You nearly convinced me removing the both length rules from our code. And yes indeed we also have smarter rules like method count and the like.

The BIGGER problem is solved as we now changed our startLine endLine implementation to give back the exact lines. Code Climate is a cloud platform that integrate with Github. When people do a push it will check their code and report on rule validations. For their UI the need to start and endline. Not a problem anymore. :-)

**Comment by up2go-rsoesemann** [comment]
@Up2Go-DavidRenz I think we should skip the length dependant rules and close this. What do you think?

**Comment by adangel** [comment]
I was wondering, how the Java rules handle this. And yes, the ExcessiveMethodLength rule is also based on the "endLine - beginLine" calculation and has the same "comment"-problem.

However, there is another way to find "big" methods: that's the NCSS (Non-commenting sourcecode statements). Java has e.g. this rule [NcssMethodCountRule](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codesize/NcssMethodCountRule.java) which more or less counts the nodes. This doesn't depend on the line numbers and it independent of the source code formatting (statements in multiple lines or in a single line...).

@up2go-rsoesemann How did you solve the "BIGGER" problem? As far as I remember, the problem is, that not each AST node from the Jorje Parser has a location info. The nodes often just return not their location, but the location of their identifier (e.g. in case of methods). That's why the start of methods is marked not with the modifier ("public") or return type, but with the name of the method.
Another weird thing (at least, the Java grammer in PMD is done differently) is, that the modifier follows the method or in other words, the modifier is a child of the method node. In Java, you have a "wrapping" node, which has the modifier, the method declaration, and the block statement as childs...

E.g. in Java:

    MethodDeclaration (this contains the public - although the start position is not accurate here, too)
        ResultType
        MethodDeclarator (<-- this contains the name)
            FormalParameters
        Block
            BlockStatements....

In Apex:

    Method (<-- this contains the name)
        Modifier (<-- this contains the public modifier)
        BlockStatement...


I think, the difficulty is, that the nodes from Jorje don't have a consistend "endIndex" property due to the fact, that it's sometimes the location of a dependendent node.
@vazexqi Consider this example:

    public class SimpleClass {
        public void method1() {
            System.out.println("abc");
            System.out.println("abc");
            int a = 1;
            // this is a comment
        }
        private static void method2() {
        }
    }

The node for "method1" has a Location with startIndex=43 and endIndex=50. You can see, that the length is 7 characters, this is just the name. As this node has as child the method body, it must be definitely longer than 7 characters. This makes using endIndex not reliable. In this case, the endIndex should be 178 (the closing bracket), and ideally the startIndex would be 31 (that would be the "public"...

**Comment by Up2Go-DavidRenz** [comment]
@up2go-rsoesemann I guess we could kick out for example the ExcessiveMethodLengthRule and replace it later with better ones like the NcssMethodCountRule mentioned by @adangel.

The bigger problem was (or is?) the missing line information in rules like the ExcessiveParameterListRule. We DO need exact line information of the relating method to highlight the related part in the code within Code Climate later.

If we get that it would be totally fine for me.

**Comment by up2go-rsoesemann** [comment]
@Up2Go-DavidRenz I am totally fine with replacing the length dependent rule with the Ncss stuff. Let's not was time on this anymore.

@adangel I didn't really solve the BIGGER problem but I found your startLine implementation to be solid. Only the endLine is not working as you seem to use invisble methods as nextsibling. There are three hidden methods clone, clinit, init that might confuse your code.

I also don't fully understand why you take the startLine of the nextSibling as EndLine. Please see my test which shows that this doesn't work.

https://github.com/Up2Go/pmd/commit/fafc3a247044d37882992d18428aaf94b2d43e6f

**Comment by up2go-rsoesemann** [comment]
Closed as duplicate of #30
